File: frontend/src/lib/dmca.ts
import { apiFetch } from "@/lib/api/client";

export type DmcaTargetType = "blog_post" | "blog_comment" | "listing" | "listing_comment";

export interface DmcaResolvedBy {
  id: string;
  email: string;
  full_name: string | null;
}

export interface DmcaTargetPreview {
  title: string;
  deleted: boolean;
}

export interface DmcaRequestRead {
  id: string;
  claimant_name: string;
  claimant_email: string;
  target_type: DmcaTargetType;
  target_id: string;
  target_preview: DmcaTargetPreview | null;
  work_description: string;
  status: "open" | "content_removed" | "rejected";
  resolution_note: string | null;
  resolved_by: DmcaResolvedBy | null;
  created_at: string;
}

export interface DmcaListResponse {
  items: DmcaRequestRead[];
  total: number;
  next_offset: number | null;
}

export function submitDmcaRequest(input: {
  claimant_name: string;
  claimant_email: string;
  target_type: DmcaTargetType;
  target_id: string;
  work_description: string;
  good_faith_statement: boolean;
  signature: string;
}) {
  return apiFetch<DmcaRequestRead>("/dmca", { method: "POST", json: input, auth: false });
}

export function listAdminDmcaRequests(params: { status?: string; limit?: number; offset?: number }) {
  const qs = new URLSearchParams();
  if (params.status) qs.set("status", params.status);
  qs.set("limit", String(params.limit ?? 50));
  qs.set("offset", String(params.offset ?? 0));
  return apiFetch<DmcaListResponse>(`/admin/dmca?${qs.toString()}`);
}

export function resolveDmcaRequest(
  id: string,
  input: { status: "content_removed" | "rejected"; resolution_note?: string },
) {
  return apiFetch<DmcaRequestRead>(`/admin/dmca/${id}`, { method: "PATCH", json: input });
}

File: backend/app/schemas/dmca.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, EmailStr, Field, field_validator

from app.services.report_service import USER_REPORTABLE_TARGET_TYPES


class DmcaRequestCreate(BaseModel):
    claimant_name: str = Field(min_length=1, max_length=200)
    claimant_email: EmailStr
    target_type: str
    target_id: str
    work_description: str = Field(min_length=1, max_length=2000)
    good_faith_statement: bool
    signature: str = Field(min_length=1, max_length=200)

    @field_validator("target_type")
    @classmethod
    def target_type_is_valid(cls, value: str) -> str:
        if value not in USER_REPORTABLE_TARGET_TYPES:
            raise ValueError(
                f"target_type must be one of {sorted(USER_REPORTABLE_TARGET_TYPES)}."
            )
        return value

    @field_validator("good_faith_statement")
    @classmethod
    def must_affirm(cls, value: bool) -> bool:
        if not value:
            raise ValueError("The good-faith statement must be affirmed.")
        return value


class DmcaResolve(BaseModel):
    status: str
    resolution_note: str | None = Field(default=None, max_length=500)

    @field_validator("status")
    @classmethod
    def status_is_valid(cls, value: str) -> str:
        if value not in ("content_removed", "rejected"):
            raise ValueError("status must be 'content_removed' or 'rejected'.")
        return value


class DmcaResolvedBy(BaseModel):
    id: str
    email: str
    full_name: str | None


class DmcaRequestRead(BaseModel):
    id: str
    claimant_name: str
    claimant_email: str
    target_type: str
    target_id: str
    target_preview: dict | None
    work_description: str
    status: str
    resolution_note: str | None
    resolved_by: DmcaResolvedBy | None
    created_at: datetime


class DmcaListResponse(BaseModel):
    items: list[DmcaRequestRead]
    total: int
    next_offset: int | None = None
